plugin/pkg/admission/wmftoolsenforcer/admission.go

```diff
@@ -0,0 +1,142 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package wmftoolsenforcer
+
+import (
+	"io"
+	"strconv"
+	"time"
+
+	"k8s.io/kubernetes/pkg/admission"
+	"k8s.io/kubernetes/pkg/api"
+	apierrors "k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/cache"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/labels"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+func init() {
+	admission.RegisterPlugin("UidEnforcer", func(client client.Interface, config io.Reader) (admission.Interface, error) {
+		return NewUidEnforcer(client), nil
+	})
+}
+
+// plugin contains the client used by the uidenforcer admin controller
+type plugin struct {
+	*admission.Handler
+	client client.Interface
+	store  cache.Store
+}
+
+// NewSecurityContextDeny creates a new instance of the SecurityContextDeny admission controller
+func NewUidEnforcer(client client.Interface) admission.Interface {
+	store := cache.NewStore(cache.MetaNamespaceKeyFunc)
+	reflector := cache.NewReflector(
+		&cache.ListWatch{
+			ListFunc: func() (runtime.Object, error) {
+				return client.Namespaces().List(labels.Everything(), fields.Everything())
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Namespaces().Watch(labels.Everything(), fields.Everything(), options)
+			},
+		},
+		&api.Namespace{},
+		store,
+		5*time.Minute,
+	)
+	reflector.Run()
+	return &plugin{
+		Handler: admission.NewHandler(admission.Create, admission.Update),
+		client:  client,
+		store:   store,
+	}
+}
+
+// This will verify the following:
+//  - Not a namespace modification from a namespaced user for their own namespace
+//  - User object has a numeric uid
+//  - Namespace object has an annotation called RunAsUser that's an integer
+//
+// If after all this there's no SecurityContext on each Container with a RunAsUser set to the same RunAsUser, it'll be set
+func (p *plugin) Admit(a admission.Attributes) (err error) {
+	if a.GetResource() != string(api.ResourcePods) && a.GetResource() != "namespace" {
+		return nil
+	}
+
+	// HACK: This should probably be a separate AdmissionController
+	hackNamespace, ok := a.GetObject().(*api.Namespace)
```

Comment:
@erictune I already had abac of the form
   {"user":"bob",  "namespace": "bob"}

and that let bob annotate namespace without denying. I'll verify this again and file a bug once I can confirm!

I've switched to the whitelist approach anyway, since you are right that does provide me more control and is what I want! I can remove the namespace modifier aspect of it now I think.